File: week5_mst/1_connecting_points/connecting_points.cpp
#include <algorithm>
#include <iostream>
#include <iomanip>
#include <vector>
#include <cmath>

using std::vector;
using std::pair;
using std::sort;

double minimum_distance(vector<int> x, vector<int> y) {
  double result = 0.;
  //write your code here
  int n = x.size();
  vector<int> disjoint_set(n,-1);
  vector<int> rank(n,0);
  vector<pair<double,pair<int,int>>> adj;

  auto distance = [&](int i, int j){
    return sqrt((double)(x[i]-x[j])*(x[i]-x[j])+(y[i]-y[j])*(y[i]-y[j]));
  };

  auto edges = [&](){
    for(int i=0;i<n;++i){
      for(int j=i+1;j<n;++j)
        adj.push_back({distance(i,j),{i,j}});
    }
  };

  auto ds_find = [&](int x){
    while(disjoint_set[x]!=-1){
      x = disjoint_set[x];
    }
    return x;
  };

  auto ds_union = [&](int x,int y){
    int px = ds_find(x);
    int py = ds_find(y);
    if(px<py){
      disjoint_set[px] = py;
      rank[py] += rank[px];
    }
    else{
      disjoint_set[py] = px;
      rank[px] += rank[py];
    }
  };

  edges();
  sort(adj.begin(),adj.end());
  for(auto seg : adj){
    double d = seg.first; int i = seg.second.first; int j = seg.second.second;
    if(ds_find(i)!=ds_find(j)){
      result += d;
      ds_union(i,j);
    }
  }
  return result;
} 

int main() {
  size_t n;
  std::cin >> n;
  vector<int> x(n), y(n);
  for (size_t i = 0; i < n; i++) {
    std::cin >> x[i] >> y[i];
  }
  std::cout << std::setprecision(10) << minimum_distance(x, y) << std::endl;
}

File: week5_mst/1_connecting_points/connecting_points.py
#Uses python3
import sys
import math

def edges(x,y):
    n = len(x)
    adj = []
    f = lambda i,j: math.sqrt((x[i]-x[j])**2+(y[i]-y[j])**2)
    for i in range(n):
        for j in range(i+1,n):
            adj.append((f(i,j),i,j))
    return adj

def find(disjoint_set,x):
    while disjoint_set[x]!=x:
        x = disjoint_set[x]
    return x

def union(disjoint_set,rank,x,y):
    px = find(disjoint_set,x)
    py = find(disjoint_set,y)
    if rank[px]<rank[py]:
        disjoint_set[px] = py
        rank[py] += rank[px]
    else:
        disjoint_set[py] = px
        rank[px] += rank[py]


def minimum_distance(x, y):
    result = 0.
    #write your code here
    n = len(x)
    adj = edges(x,y)
    adj.sort()
    disjoint_set = [i for i in range(n)]
    rank = [0]*n
    for d,i,j in adj:
        if find(disjoint_set,i)!=find(disjoint_set,j):
            result += d
            union(disjoint_set,rank,i,j)

    return result


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n = data[0]
    x = data[1::2]
    y = data[2::2]
    print("{0:.9f}".format(minimum_distance(x, y)))
